generics.go

```diff
@@ -101,18 +157,106 @@ func (pkgDefs *PackagesDefinitions) parametrizeStruct(original *TypeSpecDef, ful
 	}
 
 	parametrizedTypeSpec.TypeSpec.Type = newStructTypeDef
-
+	if genericsDefinitions[original] == nil {
+		genericsDefinitions[original] = map[string]*TypeSpecDef{}
+	}
+	genericsDefinitions[original][fullGenericForm] = parametrizedTypeSpec
 	return parametrizedTypeSpec
 }
 
-func resolveType(expr ast.Expr, field *ast.Field, genericParamTypeDefs map[string]*TypeSpecDef) ast.Expr {
+// splitStructName splits a generic struct name in his parts
+func splitStructName(fullGenericForm string) (string, []string) {
+	// split only at the first '[' and remove the last ']'
+	genericParams := strings.SplitN(strings.TrimSpace(fullGenericForm)[:len(fullGenericForm)-1], "[", 2)
+	if len(genericParams) == 1 {
+		return "", nil
+	}
+
+	// generic type name
+	genericTypeName := genericParams[0]
+
+	// generic params
+	insideBrackets := 0
+	lastParam := ""
+	params := strings.Split(genericParams[1], ",")
+	genericParams = []string{}
+	for _, p := range params {
+		numOpened := strings.Count(p, "[")
+		numClosed := strings.Count(p, "]")
+		if numOpened == numClosed && insideBrackets == 0 {
+			genericParams = append(genericParams, strings.TrimSpace(p))
+			continue
+		}
+
+		insideBrackets += numOpened - numClosed
+		lastParam += p + ","
+
+		if insideBrackets == 0 {
+			genericParams = append(genericParams, strings.TrimSpace(strings.TrimRight(lastParam, ",")))
+			lastParam = ""
+		}
+	}
+
+	return genericTypeName, genericParams
+}
+
+func resolveType(expr ast.Expr, field *ast.Field, genericParamTypeDefs map[string]*genericTypeSpec) ast.Expr {
 	if asIdent, ok := expr.(*ast.Ident); ok {
 		if genTypeSpec, ok := genericParamTypeDefs[asIdent.Name]; ok {
-			return genTypeSpec.TypeSpec.Type
+			if genTypeSpec.ArrayDepth > 0 {
+				genTypeSpec.ArrayDepth--
+				return &ast.ArrayType{Elt: resolveType(expr, field, genericParamTypeDefs)}
+			}
+			return genTypeSpec.Type()
 		}
 	} else if asArray, ok := expr.(*ast.ArrayType); ok {
 		return &ast.ArrayType{Elt: resolveType(asArray.Elt, field, genericParamTypeDefs), Len: asArray.Len, Lbrack: asArray.Lbrack}
 	}
 
```

Comment:
Else if is hard to follow, please rewrite it with a `switch case` statement for better readability.

```go
	switch astExpr := expr.(type) {
	case *ast.Ident:
		//TODO fill the code here
	case *ast.ArrayType:
		return &ast.ArrayType{
			Elt:    resolveType(astExpr.Elt, field, genericParamTypeDefs),
			Len:    astExpr.Len,
			Lbrack: astExpr.Lbrack,
		}
	}
```